### ISSUE TITLE
[RFE] Support padding other than zero padding in ONNX Parser / TensorRT

### ISSUE THREAD

**Issue by cheneyfan** [NVIDIA/TensorRT#195]
## Description

```
ERROR: builtin_op_importers.cpp:1399 In function importPad:
[8] Assertion failed: mode == "constant" && value == 0
```

## Environment

**TensorRT Version**: 6.0.1.5
**GPU Type**: P4
**Nvidia Driver Version**: 418.87.01
**CUDA Version**: 10.0
**CUDNN Version**: 7.3
**Python Version (if applicable)**: 
**TensorFlow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Operating System + Version**: 
**Baremetal or Container (if container which image + tag)**:

**Comment by cheneyfan** [comment]
![image](https://user-images.githubusercontent.com/15644150/68205455-5c347680-0005-11ea-8556-19817fcf3e19.png)

![image](https://user-images.githubusercontent.com/15644150/68205469-648cb180-0005-11ea-8b41-2c6e8aab5f03.png)

**Comment by rmccorm4** [comment]
Hi @cheneyfan,

What happens when you add the `--explicitBatch` flag? It looks like explicit batch to me from the `1 x 6 x 256 x 256` in your image.

**Comment by cheneyfan** [comment]
hi@Ryan， it shows the same error.

[root@TENCENT64 /data/home/cheneyfan/TensorRT-6.0.1.5/bin]# ./trtexec --onnx=20191028_faceswap_v2.5_net_G.onnx --explicitBatch            
&&&& RUNNING TensorRT.trtexec # ./trtexec --onnx=20191028_faceswap_v2.5_net_G.onnx --explicitBatch
[10/06/2019-15:20:31] [I] === Model Options ===
[10/06/2019-15:20:31] [I] Format: ONNX
[10/06/2019-15:20:31] [I] Model: 20191028_faceswap_v2.5_net_G.onnx
[10/06/2019-15:20:31] [I] Output:
[10/06/2019-15:20:31] [I] === Build Options ===
[10/06/2019-15:20:31] [I] Max batch: explicit
[10/06/2019-15:20:31] [I] Workspace: 16 MB
[10/06/2019-15:20:31] [I] minTiming: 1
[10/06/2019-15:20:31] [I] avgTiming: 8
[10/06/2019-15:20:31] [I] Precision: FP32
[10/06/2019-15:20:31] [I] Calibration: 
[10/06/2019-15:20:31] [I] Safe mode: Disabled
[10/06/2019-15:20:31] [I] Save engine: 
[10/06/2019-15:20:31] [I] Load engine: 
[10/06/2019-15:20:31] [I] Inputs format: fp32:CHW
[10/06/2019-15:20:31] [I] Outputs format: fp32:CHW
[10/06/2019-15:20:31] [I] Input build shapes: model
[10/06/2019-15:20:31] [I] === System Options ===
[10/06/2019-15:20:31] [I] Device: 0
[10/06/2019-15:20:31] [I] DLACore: 
[10/06/2019-15:20:31] [I] Plugins:
[10/06/2019-15:20:31] [I] === Inference Options ===
[10/06/2019-15:20:31] [I] Batch: Explicit
[10/06/2019-15:20:31] [I] Iterations: 10 (200 ms warm up)
[10/06/2019-15:20:31] [I] Duration: 10s
[10/06/2019-15:20:31] [I] Sleep time: 0ms
[10/06/2019-15:20:31] [I] Streams: 1
[10/06/2019-15:20:31] [I] Spin-wait: Disabled
[10/06/2019-15:20:31] [I] Multithreading: Enabled
[10/06/2019-15:20:31] [I] CUDA Graph: Disabled
[10/06/2019-15:20:31] [I] Skip inference: Disabled
[10/06/2019-15:20:31] [I] === Reporting Options ===
[10/06/2019-15:20:31] [I] Verbose: Disabled
[10/06/2019-15:20:31] [I] Averages: 10 inferences
[10/06/2019-15:20:31] [I] Percentile: 99
[10/06/2019-15:20:31] [I] Dump output: Disabled
[10/06/2019-15:20:31] [I] Profile: Disabled
[10/06/2019-15:20:31] [I] Export timing to JSON file: 
[10/06/2019-15:20:31] [I] Export profile to JSON file: 
[10/06/2019-15:20:31] [I] 
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:604] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 730010216
----------------------------------------------------------------
Input filename:   20191028_faceswap_v2.5_net_G.onnx
ONNX IR version:  0.0.4
Opset version:    9
Producer name:    pytorch
Producer version: 1.1
Domain:           
Model version:    0
Doc string:       
----------------------------------------------------------------
WARNING: ONNX model has a newer ir_version (0.0.4) than this parser was built against (0.0.3).
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:604] Reading dangerously large protocol message.  If the message turns out to be larger than 2147483647 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:81] The total number of bytes read was 730010216
While parsing node number 0 [Pad]:
ERROR: builtin_op_importers.cpp:1399 In function importPad:
[8] Assertion failed: mode == "constant" && value == 0
[10/06/2019-15:20:32] [E] Failed to parse onnx file
[10/06/2019-15:20:32] [E] Parsing model failed
[10/06/2019-15:20:32] [E] Engine could not be created
&&&& FAILED TensorRT.trtexec # ./trtexec --onnx=20191028_faceswap_v2.5_net_G.onnx --explicitBatch

**Comment by rmccorm4** [comment]
@cheneyfan can you share your ONNX model so I can further debug this?

**Comment by dhkim0225** [comment]
Any updates?

I have a same problem, but cannot share model since there's security issue.

Even onnx.check_model() gives me a same error.

**Comment by rmccorm4** [comment]
Can you create a dummy model with the same Pad ops or whichever one causes the error?

**Comment by dhkim0225** [comment]
@rmccorm4 Long time no see :p
Sorry, I'm too busy these days that I don't have much time to make a sample. 
Pleas understand me. But I can guess when this issue happens. 

When I converted tf.pad operation to opset11 using [tf2onnx master branch - f487e5a](https://github.com/onnx/tensorflow-onnx/tree/f487e5a/tf2onnx), I could see pads operation through netron. 
After I changed opset to 10 from 11 everything was okay. 

If you don't mind, can you experiment and share the results?

Regards,

**Comment by rmccorm4** [comment]
Marking as RFE for more padding support. If you have specific use cases / model types that require this, please add more details here to support it.

**Comment by shuo-zhao** [comment]
> Marking as RFE for more padding support. If you have specific use cases / model types that require this, please add more details here to support it.

Also, the padding layer in trt is not support for padding on channel demensions, but it is a common sense in tensorflow. When convert tensorflow model to trt, it becomes a big problem to remove channel padding layers.

**Comment by shuo-zhao** [comment]
@rmccorm4

**Comment by rajeevsrao** [comment]
@OtisZZ your RFE looks similar to https://github.com/onnx/onnx-tensorrt/issues/378. Can we please move the discussion over to that?
